src/btf.cpp

```diff
@@ -0,0 +1,216 @@
+    return NULL;
+
+  struct stat st;
+
+  if (fstat(fd, &st))
+  {
+    close(fd);
+    return NULL;
+  }
+
+  unsigned char *data;
+  unsigned int size;
+
+  size = st.st_size;
+
+  data = (unsigned char *) malloc(size);
+  if (!data)
+  {
+    close(fd);
+    return NULL;
+  }
+
+  ret = read(fd, data, size);
+  if (ret <= 0)
+  {
+    close(fd);
+    free(data);
+    return NULL;
+  }
+
+  close(fd);
+
+  *sizep = size;
+  return data;
+}
+
+static struct btf* btf_raw(char *file)
+{
+  unsigned char *data;
+  ssize_t size;
+  struct btf *btf;
+
+  data = get_data(file, &size);
+  if (data == NULL)
+  {
+    std::cerr << "BTF: failed to read data from: " << file << std::endl;
+    return NULL;
+  }
+
+  btf = btf__new(data, (__u32) size);
+  free(data);
+  return btf;
+}
+
+static int libbpf_print(enum libbpf_print_level level, const char *msg, va_list ap)
+{
+  fprintf(stderr, "BTF: (%d) ", level);
+  return vfprintf(stderr, msg, ap);
+}
+
+BTF::BTF(void) : btf(NULL), state(NODATA)
+{
+  char err_buf[256];
+  bool free_path = false;
+  char *path;
+
+  // Try to get BTF file from BPFTRACE_BTF_TEST env
+  path = std::getenv("BPFTRACE_BTF_TEST");
+  if (!path)
+  {
+    struct utsname uts;
+
+    if (uname(&uts))
+    {
+      std::cerr << "BTF: failed to get uname" << std::endl;
+      return;
+    }
+
+    asprintf(&path, "/lib/modules/%s/build/vmlinux", uts.release);
+    if (!path)
+    {
+      std::cerr << "BTF: failed to get BTF path" << std::endl;
+      return;
+    }
+
+    free_path = true;
+    btf = btf__parse_elf(path, NULL);
+  }
+  else
+  {
+    btf = btf_raw(path);
+  }
+
+  int err = libbpf_get_error(btf);
+  if (err)
+  {
+    libbpf_strerror(err, err_buf, sizeof(err_buf));
+    std::cerr << "BTF: failed to read data (" << err_buf << ") from: " << path << std::endl;
+  }
+  else
+  {
+    libbpf_set_print(libbpf_print);
+    state = OK;
+  }
+
+  if (free_path)
+    free(path);
+}
+
+BTF::~BTF()
+{
+  btf__free(btf);
+}
+
+static void dump_printf(void *ctx, const char *fmt, va_list args)
+{
+  std::string *ret = static_cast<std::string*>(ctx);
+  char *str;
+
+  if (vasprintf(&str, fmt, args) < 0)
+    return;
+
+  *ret += str;
+  free(str);
+}
+
+static const char *btf_str(const struct btf *btf, __u32 off)
+{
+  if (!off)
+    return "(anon)";
+
+  return btf__name_by_offset(btf, off) ? : "(invalid)";
+}
+
+std::string BTF::c_def(std::set<std::string>& set)
+{
+  std::string ret = std::string("");
+  struct btf_dump_opts opts = { .ctx = &ret, };
+  struct btf_dump *dump;
+  char err_buf[256];
+  int err;
+
+  dump = btf_dump__new(btf, NULL, &opts, dump_printf);
+  err = libbpf_get_error(dump);
+  if (err)
+  {
+      libbpf_strerror(err, err_buf, sizeof(err_buf));
+      std::cerr << "BTF: failed to initialize dump (" << err_buf << ")" << std::endl;
+      return std::string("");
+  }
+
+  std::set<std::string> myset(set);
+  __s32 id, max = (__s32) btf__get_nr_types(btf);
+
+  for (id = 1; id < max && myset.size(); id++)
```

Comment:
id <= max

Reply:
ugh.. should have checked bpf sources.. strange, fixed